PP-480: Job Equivalence Class Optimization

bhroam opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-480](https://pbspro.atlassian.net/browse/PP-480)**

#### Problem description
* This RFE implements an optimization which will group like jobs together to simplify the scheduler's work in determining if a job can run.  Once one like job can't run, the rest of them can't either.

#### Solution description
There are two sides to this RFE.  
The first creates a new data structure called a resresv_set.  We create a resresv_set by first looking if one exists for a job.  If one does, we add the job to the resresv_set.  If one does not, we create one.  These are containers for like jobs.
The second half is when we start considering jobs to run.  If a job can't run, we will mark its resresv_set as such.  When we consider a job to run, we look to see if its resresv_set has been marked as cannot run.  If it is marked as cannot run, we don't need to search any further.  We know the job cannot run now.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

anamikau commented on test/tests/functional/pbs_equiv_classes.py:
```diff
@@ -0,0 +1,1052 @@
+    def test_proj(self):
+        """
+        Test to see that jobs from different projects fall into the same
+        equivalence class without group limits set
```
you mean without project limits set

anamikau commented on test/tests/functional/pbs_equiv_classes.py:
```diff
@@ -0,0 +1,1052 @@
+    def test_proj_server(self):
+        """
+        Test to see that jobs from different projects fall into different
+        equivalence class server project limits set
```
typo "with server project limits set"

bhroam commented on test/tests/functional/pbs_equiv_classes.py:
```diff
@@ -0,0 +1,1052 @@
+    def test_reslist1(self):
+        """
+        Test to see if jobs with resources in Resource_List that are not in
+        the sched_config resources line fall into the same equivalence class
```
The tests are similar but not the same.  test_select tests resources within the select statement.  test_reslist1 tests resources outside of the select statement requested with qsub -l.  In the code there is a distinct difference between the two, so I wrote two tests.

bhroam commented on test/tests/functional/pbs_equiv_classes.py:
```diff
@@ -0,0 +1,1052 @@
+             'Resource_List.place': 'excl'}
+        jids2 = self.submit_jobs(3, a)
+
+        self.server.manager(MGR_CMD_SET, MGR_OBJ_SERVER,
```
I hadn't really noticed the difference.  Both work.  I did a grep through the tests and didn't find any other that used the MGR_OBJ versions, so you're right.  I removed the MGR_OBJ from all of the manager() calls.

scc138 commented:
Active review/comments are taking place on the mailing list here:  http://community.pbspro.org/t/pp-480-job-equivalence-class-optimization/376

altair4 commented:
-   the first thing I encounter in the pull request diff is a doxygen
    comment saying `convert an old limit attribute name to the   new one`
    but it doesn&rsquo;t explain what conversion is taking place and why it&rsquo;s
    happening
-   what&rsquo;s `SCHED_ATR_only_explicit_psets` and why is it being removed?
-   how do the `RETURN_ALL_ERR` and `IGNORE_EQUIV_CLASS` flags interact?
    That is, does `RETURN_ALL_ERR` dominate and result in `IGNORE_EQUIV_CLASS`
    being ignored?  And wouldn&rsquo;t it be more consistent to name the new one
    `IGNORE_EQUIV_CLASS_ERR`?
-   `data_types.h`, line 1008:  *vice versa*, not *vise versa*
-   `data_types.h`, line 1013:  `julian` should be `Julian`
-   `data_types.h`, line 1015:  is `acuall` intended to be `actual`?
-   `fifo.c`, line 760:  why is `sbj` initialized but `cal_rc` isn&rsquo;t?  Ah, I
    see why:  it&rsquo;s defined twice, once at the top of `main_sched_loop()`
    and again in the `else if (rc != SUCCESS && rc != RUN_FAILURE` block.
    Remove the first instance.
-   `fifo.c`, line 971:  what does it mean for `ec_index` to be `UNSPECIFIED`?
-   `fifo.c`,  line 1170:  why was the `if (job->can_never_run [...]` code removed?
-   `fifo.c`, lines 2035, ff:  why is the return value of `find_runnable_resresv()`
    changed?
-   `fifo.c`, line 2036:  the new `start_index` parameter is undocumented
-   `fifo.c`, lines 2272-2275:  braces appear to be indented about two tabs too much
-   `job_info.c`, line 1899:  should be `new_resresv_set(void)`
-   `data_types.h`, line 784:  there should be doxygen documentation here of
    what a `resresv_set` is
-   `job_info.c`, lines 1909-1919:  it would be more future-proof to not do
    individual setting of each `resresv_set` member but rather simply call
    `calloc()` instead of `malloc()` at line 1903
-   `job_info.c`, lines 1980, 1985, 1990:  I&rsquo;ve got to ask - why is `string_dup()`
    not simply a check for/return `NULL`, else `return(strdup(str));`?  Or at worst,
    it could check the `strdup()` return value and call `log_err()` if it must.
    Insisting on doing this yourself means any efficiencies built into the
    supplied C library would not be used.
-   `job_info.c`, lines 2034, 2296:  I see uses of both `void **` and `void**`;  uses
    of the latter are almost all (20+) confined to the scheduler;  it would be
    nice to fix all of them - there **is** supposed to be a space after the variable
    type
-   `job_info.c`, lines 2036, 2315, change &ldquo;malloc( &rdquo; to &ldquo;malloc(&rdquo;
-   `job_info.c`, line 2131:  the return value type should be on a separate
    line (as the type for `create_resresv_set_by_resresv()` that&rsquo;s immediately
    below it is);  also, the brace should be in column 1
-   `job_info.c`, line 2296:  there should be a space between the cast and its
    variable being cast
-   `job_info.c`, line 2291, ff.:  spaces after `if`, `for`, and areound the `=` and `+`
    operators
-   `job_info.c`, lines 2283, 2305:  it seems slightly better to me if the
    intiialization of both `i` and `j` would occur in the `for` statement
-   `job_info.c`, lines 2333, ff:  does the fact that there are now a number of
    sets to deal with (hence multiple, smaller requests) cause any concern that
    the tidying up phase might cause noticeable slowing?
-   `job_info.h`, line 358:  missing `void` in parameter list
-   `limits.c`, lines 620-623:  this seems like a classical example of a one-
    liner:  `return(convert_oldlim_to_new(a) != NULL);`
-   `misc.c`, lines 1178, ff:  the description is wrong:  you&rsquo;re not making
    a *copy of err* - you&rsquo;re making a copy of `oerr`
-   there are lots of lines with trailing white space that should be removed
-   `node_info.c`, lines 3678-3708:  why are you continuing to run tests and
    possibly repeatedly set `not_match` to 1 after the first time?  Why not
    return immediately?
-   `node_info.c`, lines 4003-4006:  why set `ret` to 0 instead of returning 0
    immediately?
-   `resource_resv.c`, line 1074:  `return req1->amount == req2->amount;` looks
    a bit too much like an assignment;  parenthesizing the equality test
    would make it less likely to be misread
-   `resource_resv.c`, line 1088:  the doxygen commentary should explain what
    a `NULL` value for `comparr` means;  the obvious interpretation is that if
    there are no resources needing to be compared, the result should be that
    the lists are equal.  That does not appear to be what&rsquo;s happening.
-   `resource_resv.c`, line 1103:  if both `req1` and `req2` are `NULL`, they should
    compare as equal (since they are, after all, **equal**)
-   `resource_resv.c`:  understanding the two loops would be quite a bit easier
    if the `ret*` and `cur_req*` variables were treated symmetrically - that is,
    `cur_req1` rather than `cur_req` and `ret1` instead of `ret`

bhroam commented:
@altair4 See my comments inline.  The new code is hash 3e8b8b1.

* the first thing I encounter in the pull request diff is a doxygen comment saying convert an old limit attribute name to the new one but it doesn’t explain what conversion is taking place and why it’s Happening

[bmann] This level of terseness is about the same as the rest of the limit doxygen comments.  It makes sense to me.  The function takes an old style attribute name and returns the equivalent new style attribute name.  Could you make a suggestion on what to add?

* what’s SCHED_ATR_only_explicit_psets and why is it being removed?

[bmann] I’m not sure what happened here.  The external contributor that added this forgot feature to to add to the enum on his first commit.  He added it on his second commit.  Maybe that came in after I rebased?  This should be taken care of when I rebase again.  If it isn’t, I’ll be puzzled but add it back by hand.

* how do the RETURN_ALL_ERR and IGNORE_EQUIV_CLASS flags interact? That is, does RETURN_ALL_ERR dominate and result in IGNORE_EQUIV_CLASS being ignored? And wouldn’t it be more consistent to name the new one IGNORE_EQUIV_CLASS_ERR?

[bmann] I updated the text (and the code).  They don’t really interact.  What your comment made me realize is that RETURN_ALL_ERR should trump the equivalence class feature.  In regards to equivalence classes, RETURN_ALL_ERR and IGNORE_EQUIV_CLASS do the same thing.

[bmann] I thought about your suggestion of renaming the macro and chose against it.  I think the new text in the comment better explains it.

* data_types.h, line 1008: vice versa, not vise versa
* data_types.h, line 1013: julian should be Julian
* data_types.h, line 1015: is acuall intended to be actual?
* fifo.c, line 760: why is sbj initialized but cal_rc isn’t? Ah, I see why: it’s defined twice, once at the top of main_sched_loop() and again in the else if (rc != SUCCESS && rc != RUN_FAILURE block. Remove the first instance.

[bmann] done

* fifo.c, line 971: what does it mean for ec_index to be UNSPECIFIED?

[bmann] It’s the default value.  The check is just a ‘please don’t crash in case of bug’ check.  UNSPECIFIED is defined as –1, so indexing an array by –1 is generally a bad idea.

* fifo.c, line 1170: why was the if (job->can_never_run [...] code removed?

[bmann] It was old code that needed to be removed.  The scheduler doesn’t delete jobs any more.  It used to many years ago.  Half the deletion code was removed back then.  This code lingered on.  I removed it when I noticed it.  The only reason the function is being modified at all is because of an older implementation of the feature.  I used to call update_jobs_cant_run() after any job in a set couldn’t run.  I had to abandon that direction because it broke too many things.  I left this fix in.  I can revert it if you want.

* fifo.c, lines 2035, ff: why is the return value of find_runnable_resresv() Changed?

[bmann] Same reason find_non_normal_job() returns an int.  It returns an index into an array of jobs instead of the job itself.  This function used to work by starting at the front of the array and searching for a job that wasn’t marked can_not_run.  It basically turned an O(N) operation into an O(N^2).  By stashing the index it returns, I can go back there and continue on.

* fifo.c, line 2036: the new start_index parameter is undocumented
* fifo.c, lines 2272-2275: braces appear to be indented about two tabs too much
* job_info.c, line 1899: should be new_resresv_set(void)

[bmann] done

* data_types.h, line 784: there should be doxygen documentation here of what a resresv_set is

[bmann] I added one.

* job_info.c, lines 1909-1919: it would be more future-proof to not do individual setting of each resresv_set member but rather simply call calloc() instead of malloc() at line 1903

[bmann] This is how the scheduler constructors work.  They malloc() space and individually set defaults.  If I were to reimplement the constructors now, I’d probably have done them more like you suggest.  It’s more or less too late now.  The scheduler style has been set.

* job_info.c, lines 1980, 1985, 1990: I’ve got to ask - why is string_dup() not simply a check for/return NULL, else return(strdup(str));? Or at worst, it could check the strdup() return value and call log_err() if it must. Insisting on doing this yourself means any efficiencies built into the supplied C library would not be used.

[bmann] When I originally wrote the scheduler strdup() was not part of the POSIX standard.  It is now, but the use of string_dup() is used throughout the scheduler.  It does have one nice side effect which is if the malloc() fails, it uses our logging method to log the failure.

* job_info.c, lines 2034, 2296: I see uses of both void ** and void**; uses of the latter are almost all (20+) confined to the scheduler; it would be nice to fix all of them - there is supposed to be a space after the variable Type

[bmann] There are actually 118 instances of (void**) throughout PBS.  I chose to only fix the 23 instances in the scheduler.

* job_info.c, lines 2036, 2315, change “malloc( ” to “malloc(”
* job_info.c, line 2131: the return value type should be on a separate line (as the type for create_resresv_set_by_resresv() that’s immediately below it is); also, the brace should be in column 1
* job_info.c, line 2296: there should be a space between the cast and its variable being cast
* job_info.c, line 2291, ff.: spaces after if, for, and areound the = and + Operators

[bmann] I think done.  BTW, what does ff mean?

* job_info.c, lines 2283, 2305: it seems slightly better to me if the initialization of both i and j would occur in the for statement

[bmann] ‘i’ is initialized by its for() loop.  ‘j’ is not a for loop index variable.  It indexes into an array and incremented whenever it is used.  That’s what it is initialized at the top.

* job_info.c, lines 2333, ff: does the fact that there are now a number of sets to deal with (hence multiple, smaller requests) cause any concern that the tidying up phase might cause noticeable slowing?

[bmann] I’m not sure what you are asking.  When there are a lot of jobs, the initial allocation is big.  I allocate num_jobs sets * num_jobs per set (pointers).  This is the reason for the tidy up phase.  Doing a realloc() from a larger size to a smaller size should be pretty fast.

* job_info.h, line 358: missing void in parameter list
* limits.c, lines 620-623: this seems like a classical example of a one- liner: return(convert_oldlim_to_new(a) != NULL);

[bmann] done

* misc.c, lines 1178, ff: the description is wrong: you’re not making a copy of err - you’re making a copy of oerr

[bmann] I rephrased it.

* there are lots of lines with trailing white space that should be removed

[bmann] I didn’t notice any offhand.  If you want this cleared up, you’ll have to give me pointers to instances.

* node_info.c, lines 3678-3708: why are you continuing to run tests and possibly repeatedly set not_match to 1 after the first time? Why not return immediately?

[bmann] After reviewing this function, I didn’t see any purpose of not_match.  I just replaced not_match with a bunch of return 0s.

* node_info.c, lines 4003-4006: why set ret to 0 instead of returning 0 Immediately?

[bmann] I thought about that.  I chose to use ret because of the nested if/for.  I thought a return in the middle of that would be easily missed for someone looking how you could leave the function.  This way there is just the return at the bottom.

* resource_resv.c, line 1074: return req1->amount == req2->amount; looks a bit too much like an assignment; parenthesizing the equality test would make it less likely to be misread

[bmann] I’m not sure I 100% agree.  Why would anyone return an assignment?  That just sounds weird. I made the change though.

* resource_resv.c, line 1088: the doxygen commentary should explain what a NULL value for comparr means; the obvious interpretation is that if there are no resources needing to be compared, the result should be that the lists are equal. That does not appear to be what’s happening.
* resource_resv.c, line 1103: if both req1 and req2 are NULL, they should compare as equal (since they are, after all, equal)
* resource_resv.c: understanding the two loops would be quite a bit easier if the ret* and cur_req* variables were treated symmetrically - that is, cur_req1 rather than cur_req and ret1 instead of ret

[bmann] Done

altair4 commented:
Thanks.  I think just these remain:

-   the first thing I encounter in the pull request diff is a doxygen
    comment saying `convert an old limit attribute name to the new one`
    but it doesn&rsquo;t explain what conversion is taking place and why it&rsquo;s
    happening

    For a suggestion on what to add for the comment, how about something
    like a couple `@see` references (to the old and new attribute names)?
-   `job_info.c`, lines 2036, 2315, change &ldquo;malloc( &rdquo; to &ldquo;malloc(&rdquo;

    Missed one (now at 2316).
-   `job_info.c`, line 2296:  there should be a space between the cast and its
    variable being cast

    Missed it - now at 2297.
-   `job_info.c`, line 2291, ff.:  spaces after `if`, `for`, and around the `=` and `+`
    operators

    Missed one at 2298 (and I missed saying that there ought to be a space
    around the `*` as well).
-   `job_info.c`, lines 2333, ff:  does the fact that there are now a number of
    sets to deal with (hence multiple, smaller requests) cause any concern that
    the tidying up phase might cause noticeable slowing?

    The question is really whether the fact that `realloc()` is being
    called many times rather than once introduces a performance problem.
    Note that the reallocation from larger to smaller still introduces
    overhead because the memory allocator isn&rsquo;t prescient - it has to
    contract one allocation and free the rest for each `realloc()` call.

-   there are lots of lines with trailing white space that should be removed

    To find them, start with `egrep -rc '[ ]+$' src/scheduler/` and ignore
    the first 21  that are the fault of the license preamble (inside the
    braces there should be one space character and one tab character).

bhroam commented:
* for the limits doxygen header, I added a couple of @see's for the two old2new arrays I'm using to convert.
* made whiltespace changes in job_info.c
* Removed all whitespace from the end of lines for all files.  I used sed, so it got them all, including the ones at the end of the license lines.  I committed this as a separate commit to make it easier to code review.  For some reason sed took it upon itself to delete a couple of trailing +'s.  I made sure it didn't do anything else nefarious.  I added the +'s back in the subsequent commit.
* realloc(): In my testing, the creation of the job sets take 0 seconds even if we're creating one set per job.  I really think we need to tidy up.  The worst case scenario is when we create one job set per job.  We'll end up creating num_jobs sets each with a job array of num_jobs length.  This means we have num_jobs^2 pointers.  That grows really fast with the number of jobs we're trying to scale to.  If I'm doing my math correctly, the worst case scenario for 100,000 jobs is 9gb.  I'd rather not have that lying around for the whole cycle just because we didn't tidy up.  As it is, we're allocating all that memory for at least a little bit of time.  That's bad enough.

altair4 commented:
Thanks - you've addressed all the items on my list.

bhroam commented:
Thanks @altair4.  @PBSPro/pbspro-maintainers this feature is ready for your review.  I'd request @arungrover for the review due to changes in scheduling.

arungrover commented on src/scheduler/fifo.c:
```diff
@@ -2240,29 +2266,44 @@ next_job(status *policy, server_info *sinfo, int flag)
+				last_queue++;
+				last_job_index = 0;
+			}
+			if (ind != -1) {
```
are there chances of reading an uninitialized index when skip is set to "SKIP_NON_NORMAL_JOBS" and last_queue is equal to sinfo->num_queues?

bhroam replied:
In practice this won't happen.  Once we reach the end of the last queue, we'll return -1.  It is good to be diligent though.  I added an extra check to make sure we didn't dereference a NULL pointer (because sinfo->queues[sinfo->num_queues] is the NULL sentinel)

bhroam commented on src/scheduler/data_types.h:
```diff
@@ -339,6 +345,7 @@ struct server_info
 	 */
 	status *policy;
 	fairshare_head *fairshare;	/* root of fairshare tree */
+	resresv_set **equiv_classes;
```
I disagree with the name change.  The run only one example could use the resresv_set structure, the jobs are not equivalent.  The equiv_sets are a set of jobs that are equivalent to each other.  In run only one, they are specifically going to be different.

bhroam replied:
I understand your concern, but don't know what to do about it.  Equivalence classes require all those parameters to function properly.  Can we combine the two set idea into one structure?  Should I rename resresv_set into job_equiv_class?

bhroam commented on src/scheduler/data_types.h:
```diff
@@ -773,6 +781,27 @@ struct group_info
+ * The set is defined by a number of attributes of the resresv.  If the attributes do
+ * not matter, they won't be used and set to NULL
+ */
+struct resresv_set
```
The logic isn't with sinfo->has_user_limit, it's actually in a function that looks at that element.  I put in an @see create_resresv_set_by_resresv() as a place to look for why the members can be NULL.
I dislike the state idea.  I chose can_not_run due to its analogous bit on the resource_resv structure.  Can you see other uses of state?

bhroam commented on src/scheduler/fifo.c:
```diff
@@ -751,6 +756,7 @@ main_sched_loop(status *policy, int sd, server_info *sinfo, schd_error **rerr)
 	queue_info *qinfo;		/* ptr to queue that job is in */
 	resource_resv *njob;		/* ptr to the next job to see if it can run */
 	int rc = 0;			/* return code to the function */
+	int sbj = 1;			/* return value from should_backfill_with)job() */
```
Good catch, it's not used any more.  I removed it.

bhroam commented on src/scheduler/fifo.c:
```diff
@@ -1996,8 +2007,13 @@ find_ready_resv_job(resource_resv **resvs)
-				if (resvs[i]->resv->resv_queue !=NULL)
-					rjob = find_runnable_resresv(resvs[i]->resv->resv_queue->jobs);
+				if (resvs[i]->resv->resv_queue !=NULL) {
+					ind = find_runnable_resresv(resvs[i]->resv->resv_queue->jobs, 0);
```
I added _ind to the function names (find_non_normal_job() as well).  There is precedent to appending _ind instead of _index for these types of functions.

bhroam commented:
I've made changes, but I'll hold off submitting them until you have made your full pass over the code.

arungrover commented on src/scheduler/job_info.h:
```diff
@@ -354,6 +354,24 @@ int is_finished_job(int error);
+resresv_set *create_resresv_set_by_resresv(status *policy, server_info *sinfo, resource_resv *resresv);
+
+/* find a resresv_set with a resresv as a template */
+int find_resresv_set_by_resresv(status *policy, resresv_set **rsets, resource_resv *resresv);
```
Please add find_resresv_set() to this file or else make it static

bhroam replied:
I added it to the header file.

arungrover commented on src/scheduler/node_info.c:
```diff
@@ -3912,6 +3952,64 @@ parse_selspec(char *select_spec)
+	if (c1 == NULL || c2 == NULL)
+		return 0;
+	
+	if(strcmp(c1->str_chunk, c2->str_chunk) != 0)
```
shouldn't we be doing a return '1' when strcmp matches? It seems that if the string matches completely then doing, compare_resource_req_list is redundant.
What do you think?

bhroam replied:
Good point.  I removed the strcmp() of the original string.  You are right that it's redundant.  I chose to go this way because we don't know if there will be any data derived and stored in the chunk.  Right now it is just the sequence number, but maybe in the future there could be more.

arungrover commented on src/scheduler/node_info.h:
```diff
@@ -310,6 +310,7 @@ const char *node_state_to_str(node_info *ninfo);
  *
  */
 place *parse_placespec(char *place_str);
+int compare_place(place *pl1, place *pl2);
```
Please add a function header for compare_place and compare_selspec like all other functions have.

bhroam replied:
Added

arungrover commented on src/scheduler/queue_info.c:
```diff
@@ -380,12 +380,36 @@ query_queue_info(status *policy, struct batch_status *queue, server_info *sinfo)
+				sinfo->has_user_limit = 1;
+			if(strstr(limname, "g:") != NULL)
+				sinfo->has_grp_limit = 1;
+			if(strstr(limname, "p:") != NULL)
```
Old limits has no project limits in it. So this check is not needed.

bhroam replied:
Good point.  I replaced the check with a comment saying it isn't needed.

arungrover commented on src/scheduler/resource_resv.h:
```diff
@@ -144,6 +144,9 @@ void set_resource_req(resource_req *req, char *val);
  */
 resource_req *dup_resource_req_list(resource_req *oreq);
 
+resource_req *dup_selective_resource_req_list(resource_req *oreq, resdef **deflist);
```
Please add a comment for this function too.

arungrover commented on src/scheduler/server_info.c:
```diff
@@ -469,12 +472,35 @@ query_server_info(status *pol, struct batch_status *server)
+				sinfo->has_user_limit = 1;
+			if(strstr(limname, "g:") != NULL)
+				sinfo->has_grp_limit = 1;
+			if(strstr(limname, "p:") != NULL)
```
We can remove this, old limits had no project limits.

bhroam replied:
Same as in query_queue_inf(), I replaced the check with a comment.

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -1896,6 +1893,467 @@ translate_fail_code(schd_error *err, char *comment_msg, char *log_msg)
+		return -1;
+
+	for (i = 0; rsets[i] != NULL; i++) {
+		if (qinfo != NULL && rsets[i]->qinfo == NULL || qinfo == NULL && rsets[i]->qinfo != NULL)
```
There are two checks for queue.  The first is just a quick if one is NULL and the other is not.  If that's the case then the resresv_set doesn't match.  Once the first check has passed, either both are NULL or both are not NULL.  The second check will either pass through if both are NULL or do a strcmp() on the queue names if they aren't.

bhroam replied:
OK I see your point that only qinfo is handled correctly.  I added similar code in for user, group, and project.  I also updated the comment to say only qinfo, user, group, or project can be NULL.

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -1896,6 +1893,467 @@ translate_fail_code(schd_error *err, char *comment_msg, char *log_msg)
+				free_resresv_set_array(rsets);
+				return NULL;
+			}
+			cur_rset->resresv_arr = malloc((len + 1) * sizeof(resource_resv));
```
I look at this as the both are about equivalent.  If we go with the way you suggest, we'll constantly be reallocing space to get more when we run out.  If we go with how I've implemented it, we tidy up at the end.  It all depends where we fall between best case and worst case.  If we do what you want the best case will still cause us to realloc multiple times.  The best case is where all the jobs fall in one class.  We'd alloc once for the resresv_sets and then many times as we kept adding jobs to the set.  The worst case is num_jobs resresv_sets with one job per set.  Here we're constantly reallocing resresv_sets, but not for jobs.  If you really want me to change this, I will, but I don't see it as necessary.

arungrover replied:
My concern was that with more jobs there is a possibility that scheduler will be allocating a huge chunk of memory, OOM killer shouldn't score us high in it's because of this. I do understand that we later on come back to same amount of memory as desired so I'm okay to have it as is.

bhroam commented:
@altair4 please resign off of the code changes that @arungrover requested.  After that I'll squash.

altair4 commented:
Looking at the post-Arun version, a small number of things:
-   `data_types.h`, line 788:  the `@see` directive instructs the user to
    look at `create_resresv_set_by_resresv()` to see why some members may be
    `NULL`, and the comments for the `resresv_set` structure mark only three
    structure members (`user`, `group`, `project`) as *can be NULL*.  However,
    it&rsquo;s easy to see that `create_resresv_set_by_resresv()` also calls
    other functions that may result in `NULL` returns - `dup_selspec()`,
    `dup_place()`, and `dup_selective_resource_req_list()`.
-   various:  a number of lines in newly-edited files have trailing
    white space
-   `job_info.c`, lines 2224, 2229, 2234:  binding rules for `&&` and `||` will
    save you, but as our coding standard document advises,
    *Parenthesize liberally to reduce the burden on the reader*.
-   `node_info.c`, line 3969:  I don&rsquo;t see a comment from Arun about
    why these lines
    
        if(strcmp(c1->str_chunk, c2->str_chunk) != 0)
          	return 0;
    
    should be removed.  Is it supposed to be evident?  If so, apologies -
    why?
-   `node_info.h`, line 159:  there shouldn&rsquo;t be any white space between
    the `**` and the function name

bhroam commented:
@altair4 Line 3969 of node_info.c was removed because it was redundant.  I was comparing the full string of the chunk and then went about comparing the parsed out pieces of it.  I chose to compare the parsed out parts of the chunk since that's what we use in the scheduler to schedule jobs.  I figured it was more correct.  The full string of the chunk isn't used in scheduling.

bhroam commented:
As it turns out I only removed spaces from the ends of lines (not tabs).  There were a plethora of lines with trailing tabs.  I removed those.  I am submitting them in a separate change list to make it easier to review

bhroam commented:
@altair4 @arungrover I've fixed Carl's comments.  Please rereview.  Don't worry about the conflict.  That came from the copyright update and removing whitespace from EOL.  That'll be fixed in the final squash.

arungrover commented:
@bhroam change looks good. Please attach test logs to the ticket as well.
Thanks for making the changes!

altair4 commented:
Latest changes seem to fix all I've complained about (as long as readers of the

>  create_resresv_set_by_resresv() for reasons why members can be NULL

comment dig deeper into the functions it calls to find all the possible null returns).

bhroam commented:
I wrote the comment this way on purpose.  I wanted one place all the logic for the creation of resresv_sets to be in one place.  I don't want the comment to go stale if the code changes.  There is no way you can change the code in creation of resresv_sets without create_resresv_set_by_resresv() changing.  Are you fine with this reasoning?

altair4 commented:
As long as that understanding is understood by the readers,  :-)
yes.

bhroam commented:
I don't know why Travis failed.  The bottom of the log shows that all of the smoke tests passed.  Any ideas @subhasisb ?

subhasisb commented:
All good now - Hiren says it was probably a network glitch and a refresh got it done.

arungrover commented:
Merged!
